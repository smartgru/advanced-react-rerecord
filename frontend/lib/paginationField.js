export default function paginationField() {
  return {
    keyArgs: false, // take full control of this field
    // We write custom functions to merge and read based on the 'first' and 'skip' args
    // first = 4 items per page
    // skip = how many items to offset
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      console.log('merging...', { first, skip, existing, incoming });

      const merged = existing ? existing.slice(0) : [];
      // we do it this way because someone might visit page 2 first, so we need to pad blank spots in the array
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
    read(existing = [], { args }) {
      // When we fire off a query, Apollo will check the cache first. We control that with a read method
      const { skip, first } = args;
      console.log('reading...', { args });
      console.log(existing);
      // See if we have the items we want
      const items = existing
        .slice(args.skip, args.skip + args.first)
        // we filter for empty spots because its likely we have padded spots with nothing in them.
        .filter(x => x);
      // If there are items, and they aren't blank spots, return them.
      if (items.length) {
        console.log('We have items! Gonna serve them from the cache', {
          items,
        });
        return items;
      }
      // Otherwise this function returns undefined and it will hit the network for the items, and call merge() for us
    },
  };
}
