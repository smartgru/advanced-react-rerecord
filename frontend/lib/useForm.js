import { useState } from 'react';

export default function useForm(initial = {}) {
  const [inputs, updateInputs] = useState(initial);

  function handleChange(e) {
    updateInputs({
      ...inputs,
      [e.target.name]: e.target.value,
    });
  }

  function resetForm() {
    // oh - just use the default state passed in
    updateInputs(initial);
  }

  return {
    inputs,
    handleChange,
    resetForm,
  };
}
