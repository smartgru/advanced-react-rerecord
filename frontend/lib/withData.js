import withApollo from 'next-with-apollo';
import { ApolloClient, InMemoryCache, ApolloLink } from '@apollo/client';
import { onError } from '@apollo/link-error';
import { createUploadLink } from 'apollo-upload-client';
import { getDataFromTree } from '@apollo/react-ssr';
import { endpoint, prodEndpoint } from '../config';
import paginationField from './paginationField';

function createClient({ headers, initialState }) {
  console.log('initialState');
  console.log(initialState);
  return new ApolloClient({
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        if (networkError) console.log(`[Network error]: ${networkError}`);
      }),
      // this uses apollo-link-http under the hood, so all the options here come from that package
      createUploadLink({
        uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
        fetchOptions: {
          credentials: 'include',
        },
        // pass the headers along from this request. This enables SSR with logged in state
        headers,
      }),
    ]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            allItems: paginationField(),
          },
        },
      },
    }).restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
