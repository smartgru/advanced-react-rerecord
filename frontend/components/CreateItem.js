import React, { Component } from 'react';
import { Mutation, useMutation } from '@apollo/client';

import gql from 'graphql-tag';
import Router from 'next/router';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createItem(
      data: {
        name: $name
        description: $description
        price: $price
        image: $image
      }
    ) {
      id
    }
  }
`;

function CreateItem() {
  const { inputs, handleChange } = useForm({
    name: 'Nice Shoes',
    description: 'soo nice',
    image: '',
    price: 500,
  });

  const [createItem, { loading, error }] = useMutation(CREATE_ITEM_MUTATION, {
    variables: inputs,
  });

  return (
    <Form
      data-testid="form"
      onSubmit={async e => {
        // Stop the form from submitting
        e.preventDefault();
        // call the mutation
        const res = await createItem();
        // change them to the single item page
        Router.push({
          pathname: '/item',
          query: { id: res.data.createItem.id },
        });
      }}
    >
      <Error error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="file">
          Image
          <input
            type="file"
            id="file"
            name="image"
            placeholder="Upload an image"
            required
            onChange={handleChange}
          />
        </label>

        <label htmlFor="name">
          Name
          <input
            type="text"
            id="name"
            name="name"
            placeholder="Name"
            required
            value={inputs.name}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="Price"
            required
            value={inputs.price}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="description">
          Description
          <textarea
            id="description"
            name="description"
            placeholder="Description"
            required
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Submit</button>
      </fieldset>
    </Form>
  );
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
