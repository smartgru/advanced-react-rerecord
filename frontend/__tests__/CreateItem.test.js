import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Router from 'next/router';
import { MockedProvider } from '@apollo/react-testing';
import wait from 'waait';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { fakeItem } from '../lib/testUtils';

const item = fakeItem();
describe('<CreateItem/>', () => {
  it('renders and matches snapshot', async () => {
    const { container } = render(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('handles state updating', async () => {
    const { container } = render(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );

    await userEvent.type(screen.getByPlaceholderText('Name'), item.name);
    await userEvent.type(
      screen.getByPlaceholderText('Price'),
      item.price.toString()
    );
    await userEvent.type(
      screen.getByPlaceholderText('Description'),
      item.description
    );

    expect(screen.getByDisplayValue(item.name)).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.price.toString())).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.description)).toBeInTheDocument();
  });
  it('creates an item when the form is submitted', async () => {
    const mocks = [
      {
        request: {
          query: CREATE_ITEM_MUTATION,
          variables: {
            name: item.name,
            description: item.description,
            image: '',
            price: item.price,
          },
        },
        result: {
          data: {
            createItem: {
              ...item,
              id: 'abc123',
              __typename: 'Item',
            },
          },
        },
      },
    ];

    const { container } = render(
      <MockedProvider mocks={mocks}>
        <CreateItem />
      </MockedProvider>
    );
    await userEvent.type(screen.getByPlaceholderText('Name'), item.name);
    await userEvent.type(
      screen.getByPlaceholderText('Price'),
      item.price.toString()
    );
    await userEvent.type(
      screen.getByPlaceholderText('Description'),
      item.description
    );
    // mock the router
    Router.router = { push: jest.fn() };
    await userEvent.click(screen.getByText('Submit'));
    await wait(500);
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: '/item',
      query: { id: 'abc123' },
    });
  });
});
