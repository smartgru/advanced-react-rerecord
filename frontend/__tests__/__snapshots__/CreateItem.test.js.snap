// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 bMIGic"
    data-testid="form"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="file"
      >
        Image
        <input
          id="file"
          name="image"
          placeholder="Upload an image"
          required=""
          type="file"
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          id="name"
          name="name"
          placeholder="Name"
          required=""
          type="text"
          value="Nice Shoes"
        />
      </label>
      <label
        for="price"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="Price"
          required=""
          type="number"
          value="500"
        />
      </label>
      <label
        for="description"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder="Description"
          required=""
        >
          soo nice
        </textarea>
      </label>
      <button
        type="submit"
      >
        Submit
      </button>
    </fieldset>
  </form>
</div>
`;
